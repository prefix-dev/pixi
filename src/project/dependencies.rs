use indexmap::{Equivalent, IndexMap, IndexSet};
use rattler_conda_types::{MatchSpec, NamelessMatchSpec, PackageName};
use std::{hash::Hash, iter::once};

/// Holds a list of dependencies where for each package name there can be multiple requirements.
///
/// This is used when combining the dependencies of multiple features. Although each target can only
/// have one requirement for a given package, when combining the dependencies of multiple features
/// there can be multiple requirements for a given package.
#[derive(Default, Debug, Clone)]
pub struct Dependencies {
    map: IndexMap<PackageName, IndexSet<NamelessMatchSpec>>,
}

impl From<IndexMap<PackageName, IndexSet<NamelessMatchSpec>>> for Dependencies {
    fn from(map: IndexMap<PackageName, IndexSet<NamelessMatchSpec>>) -> Self {
        Self { map }
    }
}

impl From<IndexMap<PackageName, NamelessMatchSpec>> for Dependencies {
    fn from(map: IndexMap<PackageName, NamelessMatchSpec>) -> Self {
        Self {
            map: map
                .into_iter()
                .map(|(k, v)| (k, once(v).collect()))
                .collect(),
        }
    }
}

impl Dependencies {
    /// Adds the given spec to the list of dependencies.
    pub fn insert(&mut self, name: PackageName, spec: NamelessMatchSpec) {
        self.map.entry(name).or_default().insert(spec);
    }

    /// Adds a list of specs to the list of dependencies.
    pub fn extend(&mut self, iter: impl IntoIterator<Item = (PackageName, NamelessMatchSpec)>) {
        for (name, spec) in iter {
            self.insert(name, spec);
        }
    }

    /// Adds a list of specs to the list of dependencies overwriting any existing requirements for
    /// packages that already exist in the list of dependencies.
    pub fn extend_overwrite(
        &mut self,
        iter: impl IntoIterator<Item = (PackageName, NamelessMatchSpec)>,
    ) {
        for (name, spec) in iter {
            *self.map.entry(name).or_default() = once(spec).collect();
        }
    }

    /// Removes all requirements for the given package and returns them.
    pub fn remove<Q: ?Sized>(
        &mut self,
        name: &Q,
    ) -> Option<(PackageName, IndexSet<NamelessMatchSpec>)>
    where
        Q: Hash + Equivalent<PackageName>,
    {
        self.map.shift_remove_entry(name)
    }

    /// Combine two sets of dependencies together where the requirements of `self` are extended if
    /// the same package is also defined in `other`.
    pub fn union(&self, other: &Self) -> Self {
        let mut map = self.map.clone();
        for (name, specs) in &other.map {
            map.entry(name.clone()).or_default().extend(specs.clone())
        }
        Self { map }
    }

    /// Combines two sets of dependencies where the requirements of `self` are overwritten if the
    /// same package is also defined in `other`.
    pub fn overwrite(&self, other: &Self) -> Self {
        let mut map = self.map.clone();
        for (name, specs) in other.map.iter() {
            map.insert(name.clone(), specs.clone());
        }
        Self { map }
    }

    /// Returns an iterator over the package names and their corresponding requirements.
    pub fn iter(
        &self,
    ) -> impl DoubleEndedIterator<Item = (&PackageName, &IndexSet<NamelessMatchSpec>)> + '_ {
        self.map.iter()
    }

    /// Returns an iterator over all the requirements.
    pub fn iter_specs(
        &self,
    ) -> impl DoubleEndedIterator<Item = (&PackageName, &NamelessMatchSpec)> + '_ {
        self.map
            .iter()
            .flat_map(|(name, specs)| specs.iter().map(move |spec| (name, spec)))
    }

    /// Returns the names of all the packages that have requirements.
    pub fn names(&self) -> impl DoubleEndedIterator<Item = &PackageName> + ExactSizeIterator + '_ {
        self.map.keys()
    }

    /// Convert this instance into an iterator over the package names and their corresponding
    pub fn into_specs(self) -> impl DoubleEndedIterator<Item = (PackageName, NamelessMatchSpec)> {
        self.map
            .into_iter()
            .flat_map(|(name, specs)| specs.into_iter().map(move |spec| (name.clone(), spec)))
    }

    /// Converts this instance into an iterator of [`MatchSpec`]s.
    pub fn into_match_specs(self) -> impl DoubleEndedIterator<Item = MatchSpec> {
        self.map.into_iter().flat_map(|(name, specs)| {
            specs
                .into_iter()
                .map(move |spec| MatchSpec::from_nameless(spec, Some(name.clone())))
        })
    }
}

impl IntoIterator for Dependencies {
    type Item = (PackageName, IndexSet<NamelessMatchSpec>);
    type IntoIter = indexmap::map::IntoIter<PackageName, IndexSet<NamelessMatchSpec>>;

    fn into_iter(self) -> Self::IntoIter {
        self.map.into_iter()
    }
}
