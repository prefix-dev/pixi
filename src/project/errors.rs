use crate::Project;
use fancy_display::FancyDisplay;
use itertools::Itertools;
use miette::{Diagnostic, LabeledSpan};
use pixi_manifest::{EnvironmentName, TaskName};
use rattler_conda_types::Platform;
use std::error::Error;
use std::fmt::{Display, Formatter};
use thiserror::Error;

/// An error that occurs when data is requested for a platform that is not supported.
/// TODO: Make this error better by also explaining to the user why a certain platform was not
///  supported and with suggestions as how to fix it.
#[derive(Debug, Clone)]
pub struct UnsupportedPlatformError {
    /// Platforms supported by the environment
    pub environments_platforms: Vec<Platform>,

    /// The environment that the platform is not supported for.
    pub environment: EnvironmentName,

    /// The platform that was requested
    pub platform: Platform,
}

impl Error for UnsupportedPlatformError {}

impl Display for UnsupportedPlatformError {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match &self.environment {
            EnvironmentName::Default => {
                write!(f, "the project does not support '{}'", self.platform)
            }
            EnvironmentName::Named(name) => write!(
                f,
                "the environment '{}' does not support '{}'",
                name, self.platform
            ),
        }
    }
}

impl Diagnostic for UnsupportedPlatformError {
    fn code(&self) -> Option<Box<dyn Display + '_>> {
        Some(Box::new("unsupported-platform".to_string()))
    }

    fn help(&self) -> Option<Box<dyn Display + '_>> {
        Some(Box::new(format!(
            "supported platforms are {}",
            self.environments_platforms.iter().format(", ")
        )))
    }

    fn labels(&self) -> Option<Box<dyn Iterator<Item = LabeledSpan> + '_>> {
        None
    }
}

/// An error that occurs when a task is requested which could not be found.
/// TODO: Make this error better.
///     - Include names that might have been meant instead
///     - If the tasks is only available for a certain platform, explain that.
#[derive(Debug, Clone, Diagnostic, Error)]
#[error("the task '{0}' could not be found", task_name.fancy_display())]
pub struct UnknownTask<'p> {
    /// The project that the platform is not supported for.
    pub project: &'p Project,

    /// The environment that the platform is not supported for.
    pub environment: EnvironmentName,

    /// The platform that was requested (if any)
    pub platform: Option<Platform>,

    /// The name of the task
    pub task_name: TaskName,
}
