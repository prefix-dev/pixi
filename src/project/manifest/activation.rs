use serde::Deserialize;

#[derive(Default, Clone, Deserialize, Debug)]
#[serde(deny_unknown_fields)]
pub struct Activation {
    pub scripts: Option<Vec<String>>,
}
