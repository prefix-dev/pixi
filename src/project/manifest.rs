use crate::project::python::PyPiRequirement;
use crate::project::SpecType;
use crate::utils::spanned::PixiSpanned;
use crate::{consts, task::Task};
use ::serde::Deserialize;
use indexmap::IndexMap;
use miette::{Context, IntoDiagnostic, LabeledSpan, NamedSource, Report};
use rattler_conda_types::{Channel, NamelessMatchSpec, Platform, Version};
use rattler_virtual_packages::{Archspec, Cuda, LibC, Linux, Osx, VirtualPackage};
use serde::Deserializer;
use serde_with::de::DeserializeAsWrap;
use serde_with::{serde_as, DeserializeAs, DisplayFromStr, PickFirst};
use std::collections::HashMap;
use std::ops::Range;
use std::path::{Path, PathBuf};
use url::Url;

/// Describes the contents of a project manifest.
#[serde_as]
#[derive(Debug, Clone, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ProjectManifest {
    /// Information about the project
    pub project: ProjectMetadata,

    /// Tasks defined in the project
    #[serde(default)]
    pub tasks: HashMap<String, Task>,

    /// Additional system requirements
    #[serde(default, rename = "system-requirements")]
    pub system_requirements: SystemRequirements,

    /// The dependencies of the project.
    ///
    /// We use an [`IndexMap`] to preserve the order in which the items where defined in the
    /// manifest.
    #[serde(default)]
    #[serde_as(as = "IndexMap<_, PickFirst<(_, DisplayFromStr)>>")]
    pub dependencies: IndexMap<String, NamelessMatchSpec>,

    /// The host-dependencies of the project.
    ///
    /// We use an [`IndexMap`] to preserve the order in which the items where defined in the
    /// manifest.
    #[serde(default, rename = "host-dependencies")]
    #[serde_as(as = "Option<IndexMap<_, PickFirst<(_, DisplayFromStr)>>>")]
    pub host_dependencies: Option<IndexMap<String, NamelessMatchSpec>>,

    /// The build-dependencies of the project.
    ///
    /// We use an [`IndexMap`] to preserve the order in which the items where defined in the
    /// manifest.
    #[serde(default, rename = "build-dependencies")]
    #[serde_as(as = "Option<IndexMap<_, PickFirst<(_, DisplayFromStr)>>>")]
    pub build_dependencies: Option<IndexMap<String, NamelessMatchSpec>>,

    /// Target specific configuration.
    ///
    /// We use an [`IndexMap`] to preserve the order in which the items where defined in the
    /// manifest.
    #[serde(default)]
    pub target: IndexMap<PixiSpanned<TargetSelector>, TargetMetadata>,

    /// Environment activation information.
    ///
    /// We use an [`IndexMap`] to preserve the order in which the items where defined in the
    /// manifest.
    pub activation: Option<Activation>,

    /// Optional python requirements
    #[serde(default, rename = "pypi-dependencies")]
    pub pypi_dependencies: Option<IndexMap<rip::PackageName, PyPiRequirement>>,
}

impl ProjectManifest {
    /// Validate the
    pub fn validate(&self, source: NamedSource, root_folder: &Path) -> miette::Result<()> {
        // Check if the targets are defined for existing platforms
        for target_sel in self.target.keys() {
            match target_sel.as_ref() {
                TargetSelector::Platform(p) => {
                    if !self.project.platforms.as_ref().contains(p) {
                        return Err(create_unsupported_platform_report(
                            source,
                            target_sel.span().unwrap_or_default(),
                            p,
                        ));
                    }
                }
            }
        }

        // parse the SPDX license expression to make sure that it is a valid expression.
        if let Some(spdx_expr) = &self.project.license {
            spdx::Expression::parse(spdx_expr)
                .into_diagnostic()
                .with_context(|| {
                    format!(
                        "failed to parse the SPDX license expression '{}'",
                        spdx_expr
                    )
                })?;
        }

        let check_file_existence = |x: &Option<PathBuf>| {
            if let Some(path) = x {
                let full_path = root_folder.join(path);
                if !full_path.exists() {
                    return Err(miette::miette!(
                        "the file '{}' does not exist",
                        full_path.display()
                    ));
                }
            }
            Ok(())
        };

        check_file_existence(&self.project.license_file)?;
        check_file_existence(&self.project.readme)?;

        Ok(())
    }

    /// Get the map of dependencies for a given spec type.
    pub fn create_or_get_dependencies(
        &mut self,
        spec_type: SpecType,
    ) -> &'_ mut IndexMap<String, NamelessMatchSpec> {
        match spec_type {
            SpecType::Run => &mut self.dependencies,
            SpecType::Host => {
                if let Some(ref mut deps) = self.host_dependencies {
                    deps
                } else {
                    self.host_dependencies.insert(IndexMap::new())
                }
            }
            SpecType::Build => {
                if let Some(ref mut deps) = self.build_dependencies {
                    deps
                } else {
                    self.build_dependencies.insert(IndexMap::new())
                }
            }
        }
    }

    /// Remove dependency given a `SpecType`.
    pub fn remove_dependency(
        &mut self,
        dep: &str,
        spec_type: &SpecType,
    ) -> miette::Result<(String, NamelessMatchSpec)> {
        let dependencies = match spec_type {
            SpecType::Run => Some(&mut self.dependencies),
            SpecType::Build => self.build_dependencies.as_mut(),
            SpecType::Host => self.host_dependencies.as_mut(),
        };

        if let Some(deps) = dependencies {
            deps.shift_remove_entry(dep).ok_or(miette::miette!(
                "Couldn't find {} in [{}]",
                console::style(dep).bold(),
                console::style(spec_type.name()).bold(),
            ))
        } else {
            Err(miette::miette!(
                "[{}] doesn't exist",
                console::style(spec_type.name()).bold()
            ))
        }
    }

    /// Remove a dependency for a `Platform`.
    pub fn remove_target_dependency(
        &mut self,
        dep: &str,
        spec_type: &SpecType,
        platform: &Platform,
    ) -> miette::Result<(String, NamelessMatchSpec)> {
        let target = PixiSpanned::from(TargetSelector::Platform(*platform));
        let target_metadata = self.target.get_mut(&target).ok_or(miette::miette!(
            "Platform: {} is not configured for this project",
            console::style(platform.as_str()).bold(),
        ))?;

        let dependencies = match spec_type {
            SpecType::Run => Some(&mut target_metadata.dependencies),
            SpecType::Build => target_metadata.build_dependencies.as_mut(),
            SpecType::Host => target_metadata.host_dependencies.as_mut(),
        };

        if let Some(deps) = dependencies {
            deps.shift_remove_entry(dep).ok_or(miette::miette!(
                "Couldn't find {} in [{}]",
                console::style(dep).bold(),
                console::style(format!("target.{}.{}", platform.as_str(), spec_type.name())).bold(),
            ))
        } else {
            Err(miette::miette!(
                "[{}] doesn't exist",
                console::style(format!("target.{}.{}", platform.as_str(), spec_type.name())).bold(),
            ))
        }
    }
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub enum TargetSelector {
    // Platform specific configuration
    Platform(Platform),
    // TODO: Add minijinja coolness here.
}

struct PlatformTargetSelector;

impl<'de> DeserializeAs<'de, TargetSelector> for PlatformTargetSelector {
    fn deserialize_as<D>(deserializer: D) -> Result<TargetSelector, D::Error>
    where
        D: Deserializer<'de>,
    {
        Ok(TargetSelector::Platform(Platform::deserialize(
            deserializer,
        )?))
    }
}

impl<'de> Deserialize<'de> for TargetSelector {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        Ok(
            DeserializeAsWrap::<Self, PickFirst<(PlatformTargetSelector,)>>::deserialize(
                deserializer,
            )?
            .into_inner(),
        )
    }
}

#[serde_as]
#[derive(Debug, Clone, Default, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct TargetMetadata {
    /// Target specific dependencies
    #[serde(default)]
    #[serde_as(as = "IndexMap<_, PickFirst<(_, DisplayFromStr)>>")]
    pub dependencies: IndexMap<String, NamelessMatchSpec>,

    /// The host-dependencies of the project.
    #[serde(default, rename = "host-dependencies")]
    #[serde_as(as = "Option<IndexMap<_, PickFirst<(_, DisplayFromStr)>>>")]
    pub host_dependencies: Option<IndexMap<String, NamelessMatchSpec>>,

    /// The build-dependencies of the project.
    #[serde(default, rename = "build-dependencies")]
    #[serde_as(as = "Option<IndexMap<_, PickFirst<(_, DisplayFromStr)>>>")]
    pub build_dependencies: Option<IndexMap<String, NamelessMatchSpec>>,

    /// Additional information to activate an environment.
    #[serde(default)]
    pub activation: Option<Activation>,

    /// Target specific tasks to run in the environment
    #[serde(default)]
    pub tasks: HashMap<String, Task>,
}

/// Describes the contents of the `[package]` section of the project manifest.
#[serde_as]
#[derive(Debug, Clone, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ProjectMetadata {
    /// The name of the project
    pub name: String,

    /// The version of the project
    #[serde_as(as = "Option<DisplayFromStr>")]
    pub version: Option<Version>,

    /// An optional project description
    pub description: Option<String>,

    /// Optional authors
    #[serde(default)]
    pub authors: Vec<String>,

    /// The channels used by the project
    #[serde_as(deserialize_as = "Vec<super::serde::ChannelStr>")]
    pub channels: Vec<Channel>,

    /// The platforms this project supports
    // TODO: This is actually slightly different from the rattler_conda_types::Platform because it
    //     should not include noarch.
    pub platforms: PixiSpanned<Vec<Platform>>,

    /// The license as a valid SPDX string (e.g. MIT AND Apache-2.0)
    pub license: Option<String>,

    /// The license file (relative to the project root)
    #[serde(rename = "license-file")]
    pub license_file: Option<PathBuf>,

    /// Path to the README file of the project (relative to the project root)
    pub readme: Option<PathBuf>,

    /// URL of the project homepage
    pub homepage: Option<Url>,

    /// URL of the project source repository
    pub repository: Option<Url>,

    /// URL of the project documentation
    pub documentation: Option<Url>,
}

#[serde_as]
#[derive(Debug, Clone, Deserialize, Default)]
#[serde(deny_unknown_fields)]
pub struct SystemRequirements {
    pub windows: Option<bool>,

    pub unix: Option<bool>,

    #[serde_as(as = "Option<DisplayFromStr>")]
    pub macos: Option<Version>,

    #[serde_as(as = "Option<DisplayFromStr>")]
    pub linux: Option<Version>,

    #[serde_as(as = "Option<DisplayFromStr>")]
    pub cuda: Option<Version>,

    pub libc: Option<LibCSystemRequirement>,

    pub archspec: Option<String>,
}

impl SystemRequirements {
    pub fn virtual_packages(&self) -> Vec<VirtualPackage> {
        let mut result = Vec::new();
        if self.windows == Some(true) {
            result.push(VirtualPackage::Win);
        }
        if self.unix == Some(true) {
            result.push(VirtualPackage::Unix);
        }
        if let Some(version) = self.linux.clone() {
            result.push(VirtualPackage::Linux(Linux { version }));
        }
        if let Some(version) = self.cuda.clone() {
            result.push(VirtualPackage::Cuda(Cuda { version }));
        }
        if let Some(version) = self.macos.clone() {
            result.push(VirtualPackage::Osx(Osx { version }))
        }
        if let Some(spec) = self.archspec.clone() {
            result.push(VirtualPackage::Archspec(Archspec { spec }))
        }
        if let Some(libc) = self.libc.clone() {
            result.push(VirtualPackage::LibC(libc.into()))
        }

        result
    }
}

#[serde_as]
#[derive(Debug, Clone, Deserialize)]
#[serde(untagged)]
pub enum LibCSystemRequirement {
    /// Only a version was specified, we assume glibc.
    GlibC(#[serde_as(as = "DisplayFromStr")] Version),

    /// Specified both a family and a version.
    OtherFamily(LibCFamilyAndVersion),
}

#[serde_as]
#[derive(Debug, Clone, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct LibCFamilyAndVersion {
    /// The libc family, e.g. glibc
    pub family: Option<String>,

    /// The minimum version of the libc family
    #[serde_as(as = "DisplayFromStr")]
    pub version: Version,
}

impl From<LibCSystemRequirement> for LibC {
    fn from(value: LibCSystemRequirement) -> Self {
        match value {
            LibCSystemRequirement::GlibC(version) => LibC {
                version,
                family: String::from("glibc"),
            },
            LibCSystemRequirement::OtherFamily(libc) => libc.into(),
        }
    }
}

impl From<LibCFamilyAndVersion> for LibC {
    fn from(value: LibCFamilyAndVersion) -> Self {
        LibC {
            version: value.version,
            family: value.family.unwrap_or_else(|| String::from("glibc")),
        }
    }
}
#[derive(Default, Clone, Deserialize, Debug)]
#[serde(deny_unknown_fields)]
pub struct Activation {
    pub scripts: Option<Vec<String>>,
}

// Create an error report for using a platform that is not supported by the project.
fn create_unsupported_platform_report(
    source: NamedSource,
    span: Range<usize>,
    platform: &Platform,
) -> Report {
    miette::miette!(
        labels = vec![LabeledSpan::at(
            span,
            format!("'{}' is not a supported platform", platform)
        )],
        help = format!(
            "Add '{platform}' to the `project.platforms` array of the {} manifest.",
            consts::PROJECT_MANIFEST
        ),
        "targeting a platform that this project does not support"
    )
    .with_source_code(source)
}

#[cfg(test)]
mod test {
    use super::ProjectManifest;
    use insta::{assert_debug_snapshot, assert_display_snapshot};

    const PROJECT_BOILERPLATE: &str = r#"
        [project]
        name = "foo"
        version = "0.1.0"
        channels = []
        platforms = []
        "#;

    #[test]
    fn test_target_specific() {
        let contents = format!(
            r#"
        {PROJECT_BOILERPLATE}

        [target.win-64.dependencies]
        foo = "3.4.5"

        [target.osx-64.dependencies]
        foo = "1.2.3"
        "#
        );
        assert_debug_snapshot!(
            toml_edit::de::from_str::<ProjectManifest>(&contents).expect("parsing should succeed!")
        );
    }

    #[test]
    fn test_mapped_dependencies() {
        let contents = format!(
            r#"
            {PROJECT_BOILERPLATE}
            [dependencies]
            test_map = {{ version = ">=1.2.3", channel="conda-forge", build="py34_0" }}
            "#
        );

        assert_debug_snapshot!(
            toml_edit::de::from_str::<ProjectManifest>(&contents).expect("parsing should succeed!")
        );
    }

    #[test]
    fn test_dependency_types() {
        let contents = format!(
            r#"
            {PROJECT_BOILERPLATE}
            [dependencies]
            my-game = "1.0.0"

            [build-dependencies]
            cmake = "*"

            [host-dependencies]
            sdl2 = "*"
            "#
        );

        assert_debug_snapshot!(
            toml_edit::de::from_str::<ProjectManifest>(&contents).expect("parsing should succeed!")
        );
    }

    #[test]
    fn test_invalid_target_specific() {
        let examples = [r#"[target.foobar.dependencies]
            invalid_platform = "henk""#];

        assert_display_snapshot!(examples
            .into_iter()
            .map(
                |example| toml_edit::de::from_str::<ProjectManifest>(&format!(
                    "{PROJECT_BOILERPLATE}\n{example}"
                ))
                .unwrap_err()
                .to_string()
            )
            .collect::<Vec<_>>()
            .join("\n"))
    }

    #[test]
    fn test_activation_scripts() {
        let contents = format!(
            r#"
            {PROJECT_BOILERPLATE}
            [activation]
            scripts = [".pixi/install/setup.sh"]

            [target.win-64.activation]
            scripts = [".pixi/install/setup.ps1"]

            [target.linux-64.activation]
            scripts = [".pixi/install/setup.sh", "test"]
            "#
        );

        assert_debug_snapshot!(
            toml_edit::de::from_str::<ProjectManifest>(&contents).expect("parsing should succeed!")
        );
    }

    #[test]
    fn test_target_specific_tasks() {
        let contents = format!(
            r#"
            {PROJECT_BOILERPLATE}
            [tasks]
            test = "test multi"

            [target.win-64.tasks]
            test = "test win"

            [target.linux-64.tasks]
            test = "test linux"
            "#
        );

        assert_debug_snapshot!(
            toml_edit::de::from_str::<ProjectManifest>(&contents).expect("parsing should succeed!")
        );
    }

    #[test]
    fn test_python_dependencies() {
        let contents = format!(
            r#"
            {PROJECT_BOILERPLATE}
            [pypi-dependencies]
            foo = ">=3.12"
            bar = {{ version=">=3.12", extras=["baz"] }}
            "#
        );

        assert_debug_snapshot!(
            toml_edit::de::from_str::<ProjectManifest>(&contents).expect("parsing should succeed!")
        );
    }
}
