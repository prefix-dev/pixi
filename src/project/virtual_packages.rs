use super::manifest::{LibCSystemRequirement, SystemRequirements};
use crate::project::errors::UnsupportedPlatformError;
use crate::project::Environment;
use itertools::Itertools;
use miette::Diagnostic;
use rattler_conda_types::{GenericVirtualPackage, Platform, Version};
use rattler_virtual_packages::{
    Archspec, Cuda, DetectVirtualPackageError, LibC, Linux, Osx, VirtualPackage,
};
use std::collections::HashMap;
use thiserror::Error;

/// The default GLIBC version to use. This is used when no system requirements are specified.
pub fn default_glibc_version() -> Version {
    "2.17".parse().unwrap()
}

/// The default linux version to use. This is used when no system requirements are specified.
pub fn default_linux_version() -> Version {
    "5.10".parse().unwrap()
}

/// Returns the default Mac OS version for the specified platform. The platform must refer to a
/// MacOS platform.
pub fn default_mac_os_version(platform: Platform) -> Version {
    match platform {
        Platform::OsxArm64 => "11.0".parse().unwrap(),
        Platform::Osx64 => "10.15".parse().unwrap(),
        _ => panic!(
            "default_mac_os_version() called with non-osx platform: {}",
            platform
        ),
    }
}

/// Returns a reasonable modern set of virtual packages that should be safe enough to assume.
/// At the time of writing, this is in sync with the conda-lock set of minimal virtual packages.
/// <https://github.com/conda/conda-lock/blob/3d36688278ebf4f65281de0846701d61d6017ed2/conda_lock/virtual_package.py#L175>
///
/// The method also takes into account system requirements specified in the project manifest.
pub fn get_minimal_virtual_packages(
    platform: Platform,
    system_requirements: &SystemRequirements,
) -> Vec<VirtualPackage> {
    // TODO: How to add a default cuda requirements
    let mut virtual_packages: Vec<VirtualPackage> = vec![];

    // Match high level platforms
    if platform.is_unix() {
        virtual_packages.push(VirtualPackage::Unix);
    }
    if platform.is_linux() {
        let version = system_requirements
            .linux
            .clone()
            .unwrap_or(default_linux_version());
        virtual_packages.push(VirtualPackage::Linux(Linux { version }));

        let (family, version) = system_requirements
            .libc
            .as_ref()
            .map(LibCSystemRequirement::family_and_version)
            .map(|(family, version)| (family.to_string(), version.clone()))
            .unwrap_or(("glibc".parse().unwrap(), default_glibc_version()));
        virtual_packages.push(VirtualPackage::LibC(LibC { family, version }));
    }

    if platform.is_windows() {
        virtual_packages.push(VirtualPackage::Win);
    }

    // Add platform specific packages
    if platform.is_osx() {
        let version = system_requirements
            .macos
            .clone()
            .unwrap_or_else(|| default_mac_os_version(platform));
        virtual_packages.push(VirtualPackage::Osx(Osx { version }));
    }

    // Cuda
    if let Some(version) = system_requirements.cuda.clone() {
        virtual_packages.push(VirtualPackage::Cuda(Cuda { version }));
    }

    // Archspec is only based on the platform for now
    if let Some(spec) = Archspec::from_platform(platform) {
        virtual_packages.push(VirtualPackage::Archspec(spec));
    }

    virtual_packages
}

impl Environment<'_> {
    /// Returns the set of virtual packages to use for the specified platform. This method
    /// takes into account the system requirements specified in the project manifest.
    pub fn virtual_packages(&self, platform: Platform) -> Vec<VirtualPackage> {
        get_minimal_virtual_packages(platform, &self.system_requirements())
    }
}

/// An error that occurs when the current platform does not satisfy the minimal virtual package
/// requirements.
#[derive(Debug, Error, Diagnostic)]
pub enum VerifyCurrentPlatformError {
    #[error("The current platform does not satisfy the minimal virtual package requirements")]
    UnsupportedPlatform(#[from] Box<UnsupportedPlatformError>),

    #[error(transparent)]
    DetectionVirtualPackagesError(#[from] DetectVirtualPackageError),

    #[error("The current system has a mismatching virtual package. The project requires '{required}' to be on build '{required_build_string}' but the system has build '{local_build_string}'")]
    MismatchingBuildString {
        required: String,
        required_build_string: String,
        local_build_string: String,
    },

    #[error("The current system has a mismatching virtual package. The project requires '{required}' to be at least version '{required_version}' but the system has version '{local_version}'")]
    MismatchingVersion {
        required: String,
        required_version: Box<Version>,
        local_version: Box<Version>,
    },

    #[error("The platform you are running on should at least have the virtual package {required} on version {required_version}, build_string: {required_build_string}")]
    MissingVirtualPackage {
        required: String,
        required_version: Box<Version>,
        required_build_string: String,
    },
}

/// Verifies if the current platform satisfies the minimal virtual package requirements.
pub fn verify_current_platform_has_required_virtual_packages(
    environment: &Environment<'_>,
) -> Result<(), VerifyCurrentPlatformError> {
    let current_platform = Platform::current();

    // Is the current platform in the list of supported platforms?
    if !environment.platforms().contains(&current_platform) {
        return Err(VerifyCurrentPlatformError::from(Box::new(
            UnsupportedPlatformError {
                environments_platforms: environment.platforms().into_iter().collect_vec(),
                platform: current_platform,
                environment: environment.name().clone(),
            },
        )));
    }

    let system_virtual_packages = VirtualPackage::current()?
        .iter()
        .cloned()
        .map(GenericVirtualPackage::from)
        .map(|vpkg| (vpkg.name.clone(), vpkg))
        .collect::<HashMap<_, _>>();
    let required_pkgs = environment
        .virtual_packages(current_platform)
        .into_iter()
        .map(GenericVirtualPackage::from);

    // Check for every local minimum package if it is available and on the correct version.
    for req_pkg in required_pkgs {
        if req_pkg.name.as_normalized() == "__archspec" {
            // Skip archspec packages completely for now.
            continue;
        }

        if let Some(local_vpkg) = system_virtual_packages.get(&req_pkg.name) {
            if req_pkg.build_string != local_vpkg.build_string {
                return Err(VerifyCurrentPlatformError::MismatchingBuildString {
                    required: req_pkg.name.as_source().to_string(),
                    required_build_string: req_pkg.build_string.clone(),
                    local_build_string: local_vpkg.build_string.clone(),
                });
            }

            if req_pkg.version > local_vpkg.version {
                // This case can simply happen because the default system requirements in get_minimal_virtual_packages() is higher than required.
                return Err(VerifyCurrentPlatformError::MismatchingVersion {
                    required: req_pkg.name.as_source().to_string(),
                    required_version: Box::from(req_pkg.version),
                    local_version: Box::from(local_vpkg.version.clone()),
                });
            }
        } else {
            return Err(VerifyCurrentPlatformError::MissingVirtualPackage {
                required: req_pkg.name.as_source().to_string(),
                required_version: Box::from(req_pkg.version),
                required_build_string: req_pkg.build_string.clone(),
            });
        }
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::project::manifest::SystemRequirements;
    use insta::assert_debug_snapshot;
    use rattler_conda_types::Platform;

    // Regression test on the virtual packages so there is not accidental changes
    #[test]
    fn test_get_minimal_virtual_packages() {
        let platforms = vec![
            Platform::NoArch,
            Platform::Linux64,
            Platform::LinuxAarch64,
            Platform::LinuxPpc64le,
            Platform::Osx64,
            Platform::OsxArm64,
            Platform::Win64,
        ];

        let system_requirements = SystemRequirements::default();

        for platform in platforms {
            let packages = get_minimal_virtual_packages(platform, &system_requirements)
                .into_iter()
                .map(GenericVirtualPackage::from)
                .collect_vec();
            let snapshot_name = format!("test_get_minimal_virtual_packages.{}", platform);
            assert_debug_snapshot!(snapshot_name, packages);
        }
    }
}
