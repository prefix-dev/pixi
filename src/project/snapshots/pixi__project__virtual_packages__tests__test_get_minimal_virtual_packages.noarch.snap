---
source: src/project/virtual_packages.rs
expression: packages
---
[]
